plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+            -- We're preserving ctx, the context at entry, and we'll use that when we return.
+            -- We could also add a new type of stack frame for this.  Exactly what we do will
+            -- depend on the final interface to builtins.
+            Failure -> Failure
+            Success (Closure arg' env') heap' ->
+                let term = Apply () fun arg'
+                in case termAsPrimIterApp term of
+                     Nothing ->
+                         throw $ LMachineException "Trying to apply invalid term" term
+                         -- Was "Cannot reduce a not immediately reducible application."  This message isn't very helpful.
+                     Just (IterApp (StaticStagedBuiltinName name) spine) ->
+                         case applyEvaluateBuiltinName heap' env' name spine of
+                           ConstAppSuccess term' -> returnL ctx heap' (Closure term' funEnv)
+                           ConstAppStuck         -> returnL ctx heap' (Closure term  funEnv)
+                           -- It's arguable what the env should be here. That depends on what the built-in can return.
+                           -- Ideally it'd always return a closed term, so the environment should be irrelevant.
+                           ConstAppFailure       -> throw $ LMachineException "ConstAppFailure" term
+                           ConstAppError _err    -> throw $ LMachineException "ConstAppError" term
+
+                     Just (IterApp DynamicStagedBuiltinName{}     _    ) ->
+                         throw $ LMachineException "Dynamic builtins not supported yet" term
+
+
+-- | This is a workaround (thanks to Roman) to get things working while the dynamic builtins interface is under development.
+applyEvaluateBuiltinName :: Heap -> Environment -> BuiltinName -> [Value TyName Name ()] -> ConstAppResult
+applyEvaluateBuiltinName heap env name =
+    runIdentity . runEvaluate (const $ Identity . evalL heap env) . runQuoteT . applyBuiltinName name
+
+evalL :: Heap -> Environment -> Plain Term -> EvaluationResult
+evalL heap env term = translateResult $ computeL [] heap (Closure term env)
+-- I'm not sure if this is doing quite the right thing.  The current
+-- strategy of the dynamic interface is that it when you're evaluating
+-- a builtin, you also pass the current PLC evaluator along with its
+-- state so that the builtin can evaluate terms properly (for example,
+-- it may need to look up a variable in an environment).  It's
+-- possible that this could update the heap, but in this case we're
+-- not getting the new heap back.  I think this is harmless, but
+-- perhaps inefficient (we may have to re-evaluate something that
+-- we've already evaluated).  We could solve this by putting the heap
+-- in a suitable monad, but let's wait until the interface has
+-- stabilised first.  I think the version above is OK for simple
+-- builtins for the time being.
```

Comment:
The `State` monad does seem appropriate here. It also has the advantage that the heap is returned even in case of `Failure`.

Reply:
The evaluators used by builtins are monadic now, so I think you can actually do this.